lightning/src/util/persist.rs

```diff
@@ -194,3 +206,857 @@ where
+	/// An update could not be applied to a monitor.
+	UpdateFailed { reason: &'a str, context: &'a str },
+}
+
+impl<'a> From<KVStoreUpdatingPersisterError<'a>> for io::Error {
+	fn from(value: KVStoreUpdatingPersisterError) -> Self {
+		match value {
+			KVStoreUpdatingPersisterError::BadMonitorName { reason, context } => io::Error::new(
+				io::ErrorKind::InvalidInput,
+				format!("BadMonitorName, {}, context: {}", reason, context),
+			),
+			KVStoreUpdatingPersisterError::BadUpdateName { reason, context } => io::Error::new(
+				io::ErrorKind::InvalidInput,
+				format!("BadUpdateName, {}, context: {}", reason, context),
+			),
+			KVStoreUpdatingPersisterError::MonitorDecodeFailed { reason, context } => {
+				io::Error::new(
+					io::ErrorKind::InvalidData,
+					format!("MonitorDecodeFailed, {}, context: {}", reason, context),
+				)
+			}
+			KVStoreUpdatingPersisterError::UpdateDecodeFailed { reason, context } => {
+				io::Error::new(
+					io::ErrorKind::InvalidData,
+					format!("UpdateDecodeFailed, {}, context: {}", reason, context),
+				)
+			}
+			KVStoreUpdatingPersisterError::StorageReadFailed { reason, context } => io::Error::new(
+				io::ErrorKind::Other,
+				format!("StorageReadFailed, {}, context: {}", reason, context),
+			),
+			KVStoreUpdatingPersisterError::UpdateFailed { reason, context } => io::Error::new(
+				io::ErrorKind::InvalidData,
+				format!("UpdateFailed, {}, context: {}", reason, context),
+			),
+		}
+	}
+}
+
+struct KVStoreUpdatingPersister<K: Deref, L: Deref>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	kv: K,
+	logger: L,
+}
+
+impl<K: Deref, L: Deref> KVStoreUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	pub fn read_channelmonitors<ES: Deref + Clone, SP: Deref + Clone, B: Deref, F: Deref + Clone>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		broadcaster: &B,
+		fee_estimator: F,
+	) -> std::io::Result<
+		Vec<(
+			BlockHash,
+			ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+		)>,
+	>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let mut res = Vec::new();
+		// for each monitor...
+		for monitor_name in self.list_monitor_names()? {
+			// ...parse the monitor
+			let (bh, monitor) = self.deserialize_monitor(
+				entropy_source.clone(),
+				signer_provider.clone(),
+				&monitor_name,
+			)?;
+			// ...parse and apply the updates with an id higher than the monitor.
+			for update_name in self.list_update_names(&monitor_name)? {
+				let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
+				if update.update_id == CLOSED_CHANNEL_UPDATE_ID
+					|| update.update_id > monitor.get_latest_update_id()
+				{
+					monitor
+						.update_monitor(&update, broadcaster, fee_estimator.clone(), &self.logger)
+						.map_err(|_| KVStoreUpdatingPersisterError::UpdateFailed {
```

Comment:
I'm curious as to whether we should have a `logger` around, really. I only use it one place, which is to announce that we couldn't delete stale updates when persisting a new monitor. It seemed like the only thing to do, as `PermanentFailure`ing the write didn't seem correct (the monitor write has to succeed to get there, after all). 

So I think the options are:
-  Don't delete stale updates when persisting a new monitor. When though?
  - Never? Write your own garbage collector, pal.
  - When reading monitors (only at startup)? This adds to the startup I/O and time, I think it's better to amortize the work if possible.
- Don't log a failure to delete stale updates.
  - Seems like asking for a future GitHub issue to "not silently fail when trying to delete stale updates... we ran out of disk because we didn't know"
- Embrace the `logger` and use it more, increasing our potential verbosity.

So I think the last one is right? If that's the case, then I agree, we can and should add more statements, particularly for errors.

Reply:
Embraced the logger.